### PR TITLE
[PositionedOverlay] Add scroll support for all scroll containers

### PR DESCRIPTION
**Pull request by sophschneider**
### WHY are these changes introduced?

Part of https://github.com/Shopify/polaris-internal/issues/1246

### WHAT is this pull request doing?

Before, the `PositionedOverlay` component (used by `Popover` and `Tooltip`) only scrolled with Polaris `Scrollable` components or the window. 

This PR adds support for all scroll containers so that `Popover` and `Tooltip` will always scroll with containers, even if its not wrapped in a `Scrollable`. 

|Before|After|
|-|-|
|Before-Scroll-Popover|After-Scroll-Popover|
|before-taxes-overlay|after-taxes-overlay|

### Alternative approaches

I picked this approach since it decouples `PositionedOverlay` from `Scrollable` and doesn't require any extra setup or knowledge from consumers, it should "just work". If this approach is undesirable, we could also either:

1. Extend `Tooltip` and `Popover` to accept a `ref` to custom scrollable containers and keep automatic `Scrollable` detection
2. Export `data-polaris-scrollable` or another more generic attribute that consumers can add to their custom scroll elements. Then update `Tooltip` and `Popover` documentation to require proper scroll container setup
3. Extend the `Scrollable` component to support the scrolling features in the settings modal, refactor the settings modal to use `Scrollable`, and update `Tooltip` and `Popover` documentation to require that all scroll containers are `Scrollable`s

### How to 🎩

* https://5d559397bae39100201eedc1-fdxylxukuv.chromatic.com/?path=/story/all-components-popover--with-scroll-container

* https://admin.web.web-ya2s.sophie-schneider.us.spin.dev/store/shop1/settings/taxes
1. Go to Taxes settings
4. Click "Sort" on the `Manage sales tax collection` card
5. Scroll so the "Sort" button moves
6. Make sure popover moves with button and closes when the button leaves the viewport

### 🎩 checklist

- [x] Tested a [snapshot](https://github.com/Shopify/polaris/blob/main/documentation/Releasing.md#-snapshot-releases)
- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)

### CODE REVIEW

File: polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx
Comment (sophschneider) [inline polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: Currently this is built to add vertically scrolling containers and not horizontally scrolling containers, but horizontal scroll support can be easily added if needed.

Performance is actually better than before since `Scrollable.forNode` was repeating a DOM search for iteration of the while loop. We are now doing single traversal while loop.

**Comment by marty-Wallace** [comment]
@sophschneider could you add a few more products to your spin instance. I'd like to test out how this scrolling change affects some of the columns in the BulkEditor as well. 

`rake dev:products:create NUM=30` should do it.

**Comment by sophschneider** [comment]
@marty-Wallace done!

File: polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx
Comment (marty-Wallace) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: Ah interesting. If there are enough rows in the bulk editor to cause the page to be vertically scrollable, then the popovers are also tracking with horizontal scrolling (I'm guessing since it is the same container that scrolls). But without a vertical scrollbar, the popovers do not track horizontally.

e.g. 
- [No vertical scrollbar:](https://admin.web.web-ya2s.sophie-schneider.us.spin.dev/store/shop1/bulk?resource_name=Product&edit=description%2Cmedia%2Ctags%2Cstatus%2Cproduct_taxonomy_node_id%2Cproduct_type%2Cvendor%2Ctemplate_suffix%2Cpublications.2.published_at%2Cvariants.price%2Cvariants.inventory_on_hand_adjustment&return_to=%2Fstore%2Fshop1%2Fproducts%3FselectedView%3Dall&selectedView=all&order=title+ascending&ids=22%2C8)
- [With vertical scrollbar](https://admin.web.web-ya2s.sophie-schneider.us.spin.dev/store/shop1/bulk?resource_name=Product&edit=description%2Cmedia%2Ctags%2Cstatus%2Cproduct_taxonomy_node_id%2Cproduct_type%2Cvendor%2Ctemplate_suffix%2Cpublications.2.published_at%2Cvariants.price%2Cvariants.inventory_on_hand_adjustment&return_to=%2Fstore%2Fshop1%2Fproducts%3FselectedView%3Dall&selectedView=all&order=title+ascending&ids=22%2C8%2C2%2C17%2C25%2C5%2C29%2C26%2C20%2C4%2C28%2C13%2C18%2C6%2C31%2C9%2C11%2C30%2C1%2C23%2C27%2C15%2C21%2C19%2C24%2C7%2C14%2C10%2C16%2C12%2C3)
 

Thoughts on adding horizontal containers as well?

Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: ill add them and see what it does! Wanted a good use case for it and the bulk editor is perfect :)

**Comment by marty-Wallace** [comment]
cc: @mattkubej - this PR should fix popover scrolling in the bulk editor so the popovers track with their cells.

File: polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx
Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: Ah it looks like the `PositionedOverlay` doesn't calculate [horizontal position](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/PositionedOverlay/utilities/math.ts#L87-L92) from the scroll containers like it does with the [vertical position](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/PositionedOverlay/utilities/math.ts#L17). I'm going to have to add that logic for horizontal scrolls too

Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: @marty-Wallace good news! I got it to work for horizontal scroll for both your examples above, but theres a problem with the bulk editor where the [scroll container is registered as 100% full screen width](https://github.com/Shopify/web/blob/main/app/shared/components/BulkEditor/components/Table/Table.css#L9) behind the sticky column so it doesn't close the popover until the end of that scroll container. I can't really think of a way to fix this from polaris' end so we might have to modify the width of the bulk editor scroll container or make the sticky column z-index higher than the popover

https://github.com/Shopify/polaris/assets/20652326/6b6aed25-9d59-41d2-a7ed-f1c5753af735

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts
Comment (sophschneider) [inline polaris-react/src/components/PositionedOverlay/utilities/math.ts:124]: This closes the popover once it reaches the left or right side of the activator. This works with all alignments of the popover and I think looks nicer than closing when reaching the center of the activator

File: polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx
Comment (marty-Wallace) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: @sophschneider nice!

Any thoughts on this @lbenie or @mattkubej? 

IMO - this is an improvement over the existing behaviour and we can fix the popover showing over the sticky column in the bulk editor as a follow up?

Comment (mattkubej) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: I think we should involve @e7chun and @kristy-chen2023 in the discussion as how we should proceed seems more like a product/UX decision.

Edit: trying to kick up the spin instance as it appears down

Comment (e7chun) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: Not a big fan of how the popover goes behind the sticky column in the video. I don't mind this overall Polaris change going out to help benefit other areas in the Admin, but I would like to resolve this as an immediate follow up., as this will now be a regression

Comment (kristy-chen2023) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: This change of having popover stick to the dropdown activator is an improvement for sure. But I agree with @e7chun that this needs an immediate follow-up - having the sticky column higher z-index or any other kinds of resolution to hide the popover.

Comment (e7chun) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: Could we make the Bulk Editor fix a part of this Polaris change? The reason I ask is because Product Category is one of the most highly engaged fields, so it's a paper cut experience that would probably affect most Bulk Editor users

Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: @e7chun they can be updated at the same time but not in this PR, it would be when the next polaris version is bumped in web that includes this change

Comment (e7chun) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: Gotcha -- So 2 separate PRs but they both go out in the same polaris version? That's fine with me 👍

**Review by marty-Wallace — APPROVED** [review]
Nice work on this! 

🎩'd and looks good to me, I'll leave it up to the core operate folks to determine how to make the bulk editor change in coordination with this?

**Review by chloerice — APPROVED** [review]
Awesome, thanks for fixing this @sophschneider!! :shipit: I can ping you or the bulk editor folks who have context on the z-index change needed to fix the sticky column issue in the upgrade PR (I've got one prepped for [12.11.0](https://github.com/Shopify/web/pull/115164), but will be waiting to ship it until we get the backward compatibilty PR shipped for BulkActions/SelectAllActions)

File: polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx
Comment (chloerice) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: Since the popover staying visible when the activator is behind the sticky columns is already a bug in production and not introduced by these changes, we shouldn't block this PR from shipping now. @sophschneider

Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: @chloerice do we need it for your HoverCard PR? I was holding off on it in case there were conflicts and wanted you to merge in first

Comment (chloerice) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: I do need it, though it's not blocking. No worries about conflicts, should be minimal since we didn't make changes to the same part of the component 🙏🏽

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts
Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts:124]: going to remove this for now since I don't have a fix for the bulk editor :( we can easily add it back in a separate PR if needed though!

File: polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx
Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx:199]: Hey all! As Chloe mentioned, this is less of a regression in the bulk editor and more of a side step to the existing experience. 

When I mentioned 

> might have to modify the width of the bulk editor scroll container or make the sticky column z-index higher than the popover

above, I explored both solutions and they aren't simple or fast to do as the bulk editor table is built by rows, not columns. So to elevate the sticky column above the popover, we would have semantically change the bulk editor table which is quite a big task.

I have some other solutions for the bulk editor team such as closing the popovers on scroll, but this would have to be implemented on a popover by popover basis in web. 

If we don't get any major pushback, I'm going to go ahead an merge this to unblock other overlay work our team is doing

**Pull request merged**
